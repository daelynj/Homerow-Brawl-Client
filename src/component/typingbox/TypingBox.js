import React from "react";

function TypingBox(props) {
  function handleChange(event) {
    props.setStart(true);
    props.onChange(event.target.value);
  }

  function detectSpace(event) {
    var code = event.keyCode || event.which;
    if (code === 32 && checkWord()) {
      event.preventDefault();
    }
  }

  function checkWord() {
    if (props.value === props.word) {
      props.onChange("");
      props.setCurrentWordIndex(props.currentWordIndex + 1);
      return true;
    }
  }

  return (
    <div className="TypingBox">
      <input
        type="text"
        value={props.value}
        onChange={handleChange}
        onKeyDown={detectSpace}
      />
    </div>
  );
}

export default TypingBox;
