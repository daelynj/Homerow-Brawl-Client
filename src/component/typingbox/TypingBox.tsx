import * as React from "react";

interface Props {
  value: string;
  onChange: (newInput: string) => void;
  checkLetter: () => void;
  word: String;
  setCurrentWordIndex: (newIndex: number) => void;
  currentWordIndex: number;
  countUp: boolean;
}

export const TypingBox = (props: Props) => {
  const handleChange = (event: any) => {
    if (props.countUp === true) {
      props.onChange(event.target.value);
    }
  };

  const detectBackspace = (event: any) => {
    var code = event.keyCode || event.which;
    if (code !== 8) {
      props.checkLetter();
    }
  };

  const detectSpace = (event: any) => {
    var code = event.keyCode || event.which;
    if (code === 32 && checkWord()) {
      event.preventDefault();
    }
  };

  const checkWord = () => {
    if (props.value === props.word) {
      props.onChange("");
      props.setCurrentWordIndex(props.currentWordIndex + 1);
      return true;
    }
  };

  return (
    <div className="TypingBox">
      <input
        type="text"
        value={props.value}
        onKeyUp={detectBackspace}
        onKeyDown={detectSpace}
        onChange={handleChange}
      />
    </div>
  );
};
