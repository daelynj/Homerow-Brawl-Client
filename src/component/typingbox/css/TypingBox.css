.box {
  padding: 1%;
  margin: 1%;
  font-size: 20px;
  width: 90%;
  border: 5px solid goldenrod;
}
