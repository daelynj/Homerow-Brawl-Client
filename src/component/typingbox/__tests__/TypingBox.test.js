import React from 'react';
import TypingBox from '../TypingBox';
import renderer from 'react-test-renderer';
import { mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe("TypingBox", () => {
  const onChange = jest.fn();
  const setStart = jest.fn();
  const onWordComplete = jest.fn().mockReturnValue(false);

  const buildProps = (newProps = {}) => ({
    value: "",
    onChange,
    onWordComplete,
    setStart,
    ...newProps,
  });

  it("renders an input field", () => {
    const TypingBoxComponent = renderer.create(<TypingBox {...buildProps()}/>).toJSON();
    expect(TypingBoxComponent).toMatchSnapshot();
  });

  it("renders the input field as empty when we pass an empty value", () => {
    const TypingBoxComponent = mount(<TypingBox {...buildProps()}/>);
    const inputValue = TypingBoxComponent.find('input').props().value;

    expect(inputValue).toEqual("");
  });

  describe("when we start typing", () => {
    const TypingBoxComponent = mount(<TypingBox {...buildProps()}/>);
    TypingBoxComponent.find("input").simulate('change', {target: {value: "t"}});
    
    it("calls start", () => {expect(setStart).toHaveBeenCalledWith(true)});
    it("calls onChange", () => {expect(onChange).toHaveBeenCalledWith("t")});
  });

  describe("when we press a space", () => {
    const TypingBoxComponent = mount(<TypingBox {...buildProps()}/>);
    TypingBoxComponent.find("input").simulate('change', {target: {value: " "}});

    it("calls onChange", () => {expect(onChange).toHaveBeenCalledWith(" ")});
  });

  describe("when we complete a word", () => {
    const onWordComplete = jest.fn().mockReturnValue(true);
    
    const TypingBoxComponent = mount(<TypingBox {...buildProps({onWordComplete: onWordComplete})}/>);
    TypingBoxComponent.find("input").simulate('keyDown', { keyCode: 32 });

    it("renders the input field as empty", () => {
      expect(TypingBoxComponent.find('input').props().value).toEqual("");
    });
  });
});
