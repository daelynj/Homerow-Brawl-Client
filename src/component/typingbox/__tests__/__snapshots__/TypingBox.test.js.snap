// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypingBox renders an input field 1`] = `
<div
  className="TypingBox"
>
  <input
    onChange={[Function]}
    onKeyDown={[Function]}
    type="text"
    value=""
  />
</div>
`;
